activerecord/lib/active_record/connection_handling.rb

```diff
@@ -304,15 +335,24 @@ def with_role(role, prevent_writes, &blk)
         end
       end
 
-      def with_shard(shard, role, prevent_writes)
-        old_shard = current_shard
+      def with_role_and_shard(role, shard, prevent_writes)
+        prevent_writes = true if role == reading_role
 
-        with_role(role, prevent_writes) do
-          self.current_shard = shard
-          yield
+        if ActiveRecord::Base.legacy_connection_handling
+          with_role(role, prevent_writes) do
+            self.role_and_shard_stack << { role: role, shard: shard, klass: self }
+            yield
+          end
+        else
+          connection_handler.while_preventing_writes(prevent_writes) do
+            self.role_and_shard_stack << { role: role, shard: shard, klass: self }
+            return_value = yield
+            return_value.load if return_value.is_a? ActiveRecord::Relation
```

Comment:
Why the connection handler need to know about relation? Not that I have a better place to put this logic but I'm curious to know why we need to care about this after this new logic.

Reply:
This is needed to get this test to pass https://github.com/rails/rails/blob/8f5f17cb62af2c616e763125547ea82e1c161ac6/activerecord/test/cases/connection_adapters/connection_handlers_multi_db_test.rb#L75-L84. We also use it in `swap_connection_handler` for the legacy version. It was implemented in https://github.com/rails/rails/pull/38339. Otherwise the return value of the connected to block won't be loaded. We had a few situations at GitHub where teams though it was loading the relation but it wasn't. This resulted in loading a relation from the primary/writer when the author meant to load from the replica/reader database.